### DISCUSSION TITLE
Recursion in RxJS

### DISCUSSION THREAD

**Discussion by Sawtaytoes** [ReactiveX/rxjs#6703] [Q&A]
I can think of a couple ways of doing recursion in RxJS and don't think either is really that intuitive to new RxJS developers.

I'm curious which you've seen and the situations you think these should be used (subject vs function):

![image](https://user-images.githubusercontent.com/3948069/144481739-b0252bd0-d040-4d93-b139-b405710acc93.png)

**Comment by voliva** [comment]
Functions all the way :)

I've had sometimes where I have a circular dependency between observables - In that case I break it out with a Subject.

But in general, I think a recursive function should be alright, I think I would do it like this:

```ts
declare function readDirContents(
  dir: string
): Array<{ filename: string; isDirectory: boolean }>;

function getFiles$(directory: string): Observable<string> {
  return from(readDirContents(directory)).pipe(
    mergeMap((file) =>
      file.isDirectory ? getFiles$(file.filename) : of(file.filename)
    )
  );
}
```

However, for this simple case you also have `expand()` operator, which does the recursive call for you:

```ts
function getFiles$(directory): Observable<string> {
  return from(readDirContents(directory)).pipe(
    expand((file) =>
      file.isDirectory ? from(readDirContents(file.filename)) : of(file)
    ),
    filter(({ isDirectory }) => !isDirectory),
    map(({ filename }) => filename)
  );
}
```

**Reply by Sawtaytoes** [reply]
I've never used the `expand` operator before. That looks like the best solution. The docs were super obtuse about it, but after testing, I get it now. A value from `expand` comes in and goes out immediately.

I'm not sure about the timing, but it looks like the callback for `expand` gets called once a value has gone through the pipeline and expand subscribes to the output value of that returned observable and immediately passes it through the pipeline, then calling its callback again until it receives a completed observable like `EMPTY` or `NEVER` correct?

---

Also, you're right, my example code wasn't recursive. I needed to the call the observable creator pipeline again, not just the `getDirectoryContents` function.
